File: CODIGO_C++/EJER2.cpp
#include <iostream>
#include <string>

using namespace std;

//Definimos nuestra clase que en este caso seria "Producto"
//Lo definiremos como private
class Producto {
//Definiremos los objetos dentro del "Private"
private:
    string nombre;
    int inventario;
    int ventasDirectas;
    int ventasPedidos;
    float precio;
//En el apartado de public, vamos a llamar la clase "Producto" del private
public:
    Producto(string nombre, int inventario, float precio) : nombre(nombre), inventario(inventario), ventasDirectas(0), ventasPedidos(0), precio(precio) {}
//Haremos un metodo de "MostrarInventario"
    void mostrarInventario() {
        cout<<"Inventario de "<<nombre<<": "<<inventario<< endl;
    }
//Haremos un metodo "venderProducto"
    void venderProducto(int cantidad) {
        if (inventario >= cantidad) {   //Se asegura de que haya suficiente stock para realizar la operación
            inventario -= cantidad;     //Si la condición anterior se cumple, se resta la cantidad especificada del inventario del producto
            //Simula la venta de esa cantidad de unidades o la realización de un pedido.
            ventasDirectas += cantidad; //Después de restar la cantidad del inventario, se incrementa el contador de ventas directas del producto en la cantidad especificada
            cout<<"Venta realizada: "<<cantidad<<" unidades de "<<nombre<< endl;
        } else {
            cout<<"No hay suficiente inventario de "<<nombre<< endl;
        }
    }

    void hacerPedido(string cliente, int cantidad) {
        if (inventario >= cantidad) {
            inventario -= cantidad;
            ventasPedidos += cantidad;  //Si la condicion anterior se cumple, se resta la cantidad del inventario del producto
            cout<<"Pedido realizado para: "<<cliente<<", Cantidad: "<<cantidad<< endl;
        } else {
            cout<<"No hay suficiente inventario de " <<nombre<< " para completar el pedido"<< endl;
        }
    }

    int getVentasDirectas() {   //Devuelve la cantidad total de ventas directas realizadas del producto
        return ventasDirectas;
    }

    int getVentasPedidos() {    //Devuelve la cantidad total de ventas por pedidos realizadas del producto.
        return ventasPedidos;
    }

    float getPrecio() {         //Devuelve el precio unitario del producto
        return precio;
    }

    void reiniciarVentas() {    //Reinicia el inventario del producto, estableciéndolo en 50 unidades, que es el valor inicial estableciéndolos en cero
        ventasDirectas = 0;
        ventasPedidos = 0;
    }

    void reiniciarInventario() {
        inventario = 50;        //Reinicia el inventario a su valor inicial
    }
};

int main() {
    Producto arroz("Arroz", 50, 8.0);       //La cantidad del producto y el precio
    Producto frijoles("Frijoles", 20, 5.0); //La cantidad del producto y el precio

    string tipoUsuario = "";
    
    while (tipoUsuario != "salir") {
        cout<<"Eres usuario, administrador o salir del programa? (usuario/administrador/salir): ";  //Opcion para ver si el usuario es usuario, admin o desea salir
        cin>>tipoUsuario;
        //Si elije que es usuario, este lo mandara a las opciones que tiene el usuario
        if (tipoUsuario == "usuario") {     
            int opcion;
            cout<<"Opciones para usuario:"<<endl;
            cout<<"1. Comprar producto"<<endl;
            cout<<"2. Hacer pedido"<<endl;
            cin>>opcion;
            //Debe ingresar la cantidad a comprar
            if (opcion == 1) {
                int cantidadCompra;
                cout<<"Productos disponibles:"<<endl;
                cout<<"1. Arroz - Precio: "<<arroz.getPrecio() << endl;
                cout<<"2. Frijoles - Precio: "<<frijoles.getPrecio() << endl;
                cout<<"Seleccione el producto a comprar (1 o 2): ";
                int productoSeleccionado;
                cin>>productoSeleccionado;
                cout<<"Ingrese la cantidad a comprar: ";
                cin>>cantidadCompra;

                if (productoSeleccionado == 1) {
                    arroz.venderProducto(cantidadCompra);
                } else if (productoSeleccionado == 2) {
                    frijoles.venderProducto(cantidadCompra);
                }
            //Si el usuario desea hacer pedido, lo mandara a la interfaz de los pedidos
            } else if (opcion == 2) {
                string clientePedido;
                int cantidadPedido;
                cout<<"Productos disponibles para pedido:"<<endl;
                cout<<"1. Arroz"<<endl;
                cout<<"2. Frijoles"<<endl;
                cout<<"Seleccione el producto a pedir (1 o 2): ";
                int productoSeleccionado;
                cin>>productoSeleccionado;
                cout<<"Ingrese su nombre para el pedido: ";
                cin>>clientePedido;
                cout<<"Ingrese la cantidad deseada: ";
                cin>>cantidadPedido;
                //Se encarga de tomar la cantidad de los productos ingresados
                if (productoSeleccionado == 1) {
                    arroz.hacerPedido(clientePedido, cantidadPedido);
                } else if (productoSeleccionado == 2) {
                    frijoles.hacerPedido(clientePedido, cantidadPedido);
                }
            }
        //En el interfaz de administrador, tendra dos opciones, ver inventario y ver ventas totales
        } else if (tipoUsuario == "administrador") {
            int opcionAdmin;
            cout<<"Opciones para administrador:" << endl;
            cout<<"1. Ver inventario"<<endl;
            cout<<"2. Ver ventas totales"<<endl;
            cin>>opcionAdmin;
            //Si elije ver inventario, podra ver la cantidad vendida de cada producto y las ganancias de estos
            //Tambien podra ver los pedidos realizados por el usuario
            if (opcionAdmin == 1) {
                arroz.mostrarInventario();
                frijoles.mostrarInventario();
            } else if (opcionAdmin == 2) {
                cout<<"Ventas totales de Arroz por compras directas: "<<arroz.getVentasDirectas()<<" - Ganancia: "<<arroz.getVentasDirectas() * arroz.getPrecio()<<endl;
                cout<<"Ventas totales de Frijoles por compras directas: "<<frijoles.getVentasDirectas()<<" - Ganancia: "<<frijoles.getVentasDirectas() * frijoles.getPrecio()<<endl;
                cout<<"Ventas totales de Arroz por pedidos: "<<arroz.getVentasPedidos() << " - Ganancia: "<<arroz.getVentasPedidos() * arroz.getPrecio()<<endl;
                cout<<"Ventas totales de Frijoles por pedidos: "<<frijoles.getVentasPedidos() << " - Ganancia: "<<frijoles.getVentasPedidos() * frijoles.getPrecio()<<endl;
                cout<<"Ventas totales de todos los pedidos: "<<(arroz.getVentasDirectas() + arroz.getVentasPedidos()) * arroz.getPrecio() + (frijoles.getVentasDirectas() + frijoles.getVentasPedidos()) * frijoles.getPrecio()<<endl;
            }
        //Si el usuario desea salir del programa, solo debera escribir "salir"
        } else if (tipoUsuario == "salir") {
            cout<<"Saliendo del programa..."<<endl;
        } else {
        //Muestra mensaje si no ingresa la opcion valida
            cout<<"Opcion no valida. Por favor, seleccione 'usuario', 'administrador' o 'salir'."<<endl;
        }
    }

    //Reinicia el inventario y las ventas de ambos productos
    arroz.reiniciarInventario();
    arroz.reiniciarVentas();
    frijoles.reiniciarInventario();
    frijoles.reiniciarVentas();

    return 0;
}

File: CODIGO_PYTHON/EJER2.py
class Producto:
    #Definiremos los objetos dentro del "private"
    def __init__(self, nombre, inventario, precio):
        self.nombre = nombre
        self.inventario = inventario
        self.ventas_directas = 0
        self.ventas_pedidos = 0
        self.precio = precio

    #Haremos un método de "mostrarinventario"
    def mostrar_inventario(self):
        print("Inventario de", self.nombre + ":", self.inventario)

    #Haremos un método "venderproducto"
    def vender_producto(self, cantidad):
        if self.inventario > cantidad:
            self.inventario -= cantidad
            self.ventas_directas += cantidad
            print("Venta realizada:", cantidad, "unidades de", self.nombre)
        else:
            print("No hay suficiente inventario de", self.nombre)

    def hacer_pedido(self, cliente, cantidad):
        if self.inventario > cantidad:
            self.inventario -= cantidad
            self.ventas_pedidos += cantidad
            print("Pedido realizado para:", cliente + ", Cantidad:", cantidad)
        else:
            print("No hay suficiente inventario de", self.nombre, "para completar el pedido")

    def get_ventas_directas(self):
        return self.ventas_directas

    def get_ventas_pedidos(self):
        return self.ventas_pedidos

    def get_precio(self):
        return self.precio

    def reiniciar_ventas(self):
        self.ventas_directas = 0
        self.ventas_pedidos = 0

    def reiniciar_inventario(self):
        self.inventario = 50

#Reinicia el inventario a su valor inicial

if __name__ == "__main__":
    arroz = Producto("Arroz", 50, 8.0) #La cantidad del producto y el precio
    frijoles = Producto("Frijoles", 20, 5.0) # La cantidad del producto y el precio
    tipo_usuario = ""
    while tipo_usuario != "salir":
        print("Eres usuario, administrador o salir del programa? (usuario/administrador/salir): ")
        tipo_usuario = input()
        if tipo_usuario == "usuario":
            print("Opciones para usuario:")
            print("1. Comprar producto")
            print("2. Hacer pedido")
            opcion = int(input())
            if opcion == 1:
                print("Productos disponibles:")
                print("1. Arroz Precio:", arroz.get_precio())
                print("2. Frijoles Precio:", frijoles.get_precio())
                print("Seleccione el producto a comprar (1 o 2):")
                producto_seleccionado = int(input())
                print("Ingrese la cantidad a comprar:")
                cantidad_compra = int(input())
                if producto_seleccionado == 1:
                    arroz.vender_producto(cantidad_compra)
                elif producto_seleccionado == 2:
                    frijoles.vender_producto(cantidad_compra)
            elif opcion == 2:
                print("Productos disponibles para pedido:")
                print("1. Arroz")
                print("2. Frijoles")
                print("Seleccione el producto a pedir (1 o 2):")
                producto_seleccionado = int(input())
                print("Ingrese su nombre para el pedido:")
                cliente_pedido = input()
                print("Ingrese la cantidad deseada:")
                cantidad_pedido = int(input())
                if producto_seleccionado == 1:
                    arroz.hacer_pedido(cliente_pedido, cantidad_pedido)
                elif producto_seleccionado == 2:
                    frijoles.hacer_pedido(cliente_pedido, cantidad_pedido)
        elif tipo_usuario == "administrador":
            print("Opciones para administrador:")
            print("1. Ver inventario")
            print("2. Ver ventas totales")
            opcion_admin = int(input())
            if opcion_admin == 1:
                arroz.mostrar_inventario()
                frijoles.mostrar_inventario()
            elif opcion_admin == 2:
                print("Ventas totales de Arroz por compras directas:", arroz.get_ventas_directas(), "Ganancia:", arroz.get_ventas_directas() * arroz.get_precio())
                print("Ventas totales de Frijoles por compras directas:", frijoles.get_ventas_directas(), "Ganancia:", frijoles.get_ventas_directas() * frijoles.get_precio())
                print("Ventas totales de Arroz por pedidos:", arroz.get_ventas_pedidos(), "Ganancia:", arroz.get_ventas_pedidos() * arroz.get_precio())
                print("Ventas totales de Frijoles por pedidos:", frijoles.get_ventas_pedidos(), "Ganancia:", frijoles.get_ventas_pedidos() * frijoles.get_precio())
        elif tipo_usuario == "salir":
            print("Saliendo del programa...")
        else:
            print("Opción no válida. Por favor, seleccione 'usuario', 'administrador' o 'salir'.")
        arroz.reiniciar_inventario()
        arroz.reiniciar_ventas()
        frijoles.reiniciar_inventario()
        frijoles.reiniciar_ventas()
